apps/typegpu-docs/src/content/examples/rendering/cubemap-reflection/icosphere.ts

```diff
@@ -0,0 +1,313 @@
+      }
+    });
+
+    this.pipeline = this.root['~unstable']
+      .withCompute(computeFn)
+      .createPipeline();
+  }
+
+  createIcosphere(subdivisions: number, smooth: boolean): IcosphereBuffer {
+    if (this.maxBufferSize) {
+      let safeSize = subdivisions;
+      while (
+        getVertexAmount(safeSize) * d.sizeOf(Vertex) >
+        this.maxBufferSize
+      ) {
+        safeSize--;
+      }
+      if (safeSize < subdivisions) {
+        console.warn(
+          `Requested icosphere of size ${getVertexAmount(subdivisions) * d.sizeOf(Vertex)} exceeds max buffer size of ${this.maxBufferSize} - reducing subdivisions to ${safeSize}`,
+        );
+        return this.createIcosphere(safeSize, smooth);
+      }
+    }
+
+    const key = `${subdivisions}-${smooth}`;
+    const cached = this.cache.get(key);
+    if (cached) {
+      return cached;
+    }
+
+    const buffer = this.subdivide(subdivisions, smooth);
+    this.cache.set(key, buffer);
+    return buffer;
+  }
+
+  private subdivide(
+    wantedSubdivisions: number,
+    smooth: boolean,
+  ): IcosphereBuffer {
+    if (wantedSubdivisions === 0) {
+      const key = `${wantedSubdivisions}-${smooth}`;
+      const cached = this.cache.get(key);
+      if (cached) {
+        return cached;
+      }
+
+      const initialVertices = this.root
+        .createBuffer(
+          d.disarrayOf(Vertex, getVertexAmount(0)),
+          createBaseIcosphere(smooth),
+        )
+        .$usage('vertex')
+        .$addFlags(GPUBufferUsage.STORAGE);
+      this.cache.set(key, initialVertices);
+      return initialVertices;
+    }
+
+    const previousKey = `${wantedSubdivisions - 1}-${smooth}`;
+    let previousVertices = this.cache.get(previousKey);
+    if (!previousVertices) {
+      previousVertices = this.subdivide(wantedSubdivisions - 1, smooth);
+    }
+
+    const nextBuffer = this.root
+      .createBuffer(d.disarrayOf(Vertex, getVertexAmount(wantedSubdivisions)))
+      .$usage('vertex')
+      .$addFlags(GPUBufferUsage.STORAGE);
+
+    const currentComputeView = this.root
+      .createBuffer(
+        d.arrayOf(ComputeVertex, getVertexAmount(wantedSubdivisions - 1)),
+        previousVertices.buffer,
+      )
+      .$usage('storage');
+    const nextComputeView = this.root
+      .createBuffer(
+        d.arrayOf(ComputeVertex, getVertexAmount(wantedSubdivisions)),
+        nextBuffer.buffer,
+      )
+      .$usage('storage');
+
+    this.smoothBuffer.write(smooth ? 1 : 0);
+
+    const bindGroup = this.root.createBindGroup(generatorLayout, {
+      prevVertices: currentComputeView,
+      nextVertices: nextComputeView,
+      smoothFlag: this.smoothBuffer,
+    });
+
+    const triangleCount = getVertexAmount(wantedSubdivisions - 1) / 3;
+    const xGroups = Math.min(triangleCount, 65535);
+    const yGroups = Math.ceil(triangleCount / 65535);
+
+    this.pipeline
+      .with(generatorLayout, bindGroup)
+      .dispatchWorkgroups(xGroups, yGroups, 1);
```

Comment:
workgroup size is set to [64, 1, 1], so aren't we invoking 64x too many threads?